common/lib/xmodule/xmodule/modulestore/mongo/base.py

```diff
@@ -511,13 +514,22 @@ def has(self, key):
 
     @contract(key=unicode, value="BlockUsageLocator | None")
     def set(self, key, value):
+        # If the key already exists, we have to remove it from the reverse lookup
+        if key in self:
+            old_val = self[key]
+            old_val_keys = self._values_to_keys[old_val]
+            if key in old_val_keys:
+                old_val_keys.remove(key)
         self[key] = value
+        self._values_to_keys[value].add(key)
 
     @contract(value="BlockUsageLocator")
     def delete_by_value(self, value):
-        keys_to_delete = [k for k, v in self.iteritems() if v == value]
+        keys_to_delete = self._values_to_keys.pop(value, set())
         for key in keys_to_delete:
-            del self[key]
+            if key in self:
+                del self[key]
+
```

Comment:
Thanks. I hadn't updated it because there was a weirdness in the coverage diff that I didn't understand. https://build.testeng.edx.org/job/edx-platform-all-tests-pr-flow/8934/Diff_Coverage_Report/

I'll look at it after-hours tonight.